### PR TITLE
Codegeneration for API mocking with MSW in frontend blocks

### PR DESCRIPTION
**Pull request by andertun**
If a frontend block consumes `'kapeta/resource-type-rest-client'` it will change som files and create mock api handlers for each endpoint

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -90,20 +93,47 @@ export default class ReactTSTarget extends Target {
...
         });
 
+        engine.registerHelper('ifVoidType', (type: TypeLike, options: HelperOptions) => {
+            // Check if type has a property 'string' (which it would have if it is a SafeString)
+            if (type && typeof type === 'object' && 'string' in type) {
+                type = type.string as string;
+            }
+
+            let isVoid = false;
+            if (typeof type === 'string') {
```
Comment (hofmeister) [inline src/index.ts:113]: there is a "typeName(type)" helper which will do this for you

File: src/index.ts [hunk 2]
```diff
@@ -90,20 +93,47 @@ export default class ReactTSTarget extends Target {
         });
 
         engine.registerHelper('ifValueType', (type, options) => {
-            if ((type?.type || type?.ref) && type?.type?.toLowerCase() !== 'void' && type?.ref?.toLowerCase() !== 'void') {
+            if (
+                (type?.type || type?.ref) &&
```
Comment (hofmeister) [inline src/index.ts:97]: Use typeName(type) helper

**Comment by hofmeister** [comment]
the "typeName" helper is in @kapeta/schemas btw

File: templates/kapeta/block-type-frontend/tsconfig.json [hunk 3]
```diff
@@ -6,12 +6,12 @@
     "skipLibCheck": true,
     "esModuleInterop": true,
     "strict": true,
-    "module": "CommonJS",
+    "module": "ESNext",
```
[the remark this answers is not in the document]
Comment (andertun) [reply templates/kapeta/block-type-frontend/tsconfig.json]: The reason was we now have a dynamic import `import ('./browser')` which TS complained about. Changing `module` and `moduleResolution` solved that but I forgot that the same `tsconfig.json` is also used in the server part of the frontend block.

I have now changed it back and made a workaround here c5ea0021a32cc02698ab6063d7a1eeb819606e47. 

I think we probably could benefit from a better division of server/browser code so we can have a base tsconfig in the root with common settings and then a tsconfig in each that extends from it.

File: src/index.ts (hunk 1 above)
Comment (andertun) [reply src/index.ts]: 740ae9bb97f26d05a86cdd8058af83160c2e52a2 Here is an updated version where I use that version... but I'm not sure it is much cleaner. The `TypeLike` in codegen-target is not the same type as `TypeLike` in schemas.

```
export type TypeLike = string | {
    type?: string;
    ref?: string;
};
```
vs
```
export interface TypeLike {
    type?: string;
    ref?: string;
}
```

File: src/index.ts [hunk 4]
```diff
@@ -89,21 +93,43 @@ export default class ReactTSTarget extends Target {
...
 
+        engine.registerHelper('ifVoidType', (type: TypeLike, options: HelperOptions) => {
+            // Check if type has a property 'string' (which it would have if it is a SafeString)
+            if (type && typeof type === 'object' && 'string' in type) {
+                type = type.string as string;
+            }
+
+            if (
+                (typeof type === 'string' && type === 'void') ||
+                (typeof type !== 'string' && typeName(type) === 'void')
```
Comment (hofmeister) [inline src/index.ts:114]: typeName will actually accept a "TypeLike" argument - you shouldn't need the typeof checks

**Review by hofmeister — APPROVED** [review]
Minor thing - otherwise 👍

File: src/index.ts (hunk 4 above)
Comment (andertun) [reply src/index.ts:114]: The `TypeLike` in codegen-target is not the same type as `TypeLike` in schemas. The `TypeLike` type the typeName function expects is not the same as is used in the codegen-target. 

```
export type TypeLike = string | {
    type?: string;
    ref?: string;
};
```
vs
```
export interface TypeLike {
    type?: string;
    ref?: string;
}
```

**Pull request merged**